Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/local/storage/LocalStatisticsConfigStorage.kt

```diff
@@ -2,49 +2,65 @@ package de.rki.coronawarnapp.statistics.local.storage
 
 import android.content.Context
 import com.google.gson.Gson
+import de.rki.coronawarnapp.datadonation.analytics.common.federalStateShortName
 import de.rki.coronawarnapp.statistics.local.FederalStateToPackageId
 import de.rki.coronawarnapp.util.di.AppContext
 import de.rki.coronawarnapp.util.preferences.FlowPreference
 import de.rki.coronawarnapp.util.preferences.createFlowPreference
 import de.rki.coronawarnapp.util.serialization.BaseGson
+import de.rki.coronawarnapp.util.serialization.adapter.RuntimeTypeAdapterFactory
 import kotlinx.coroutines.flow.map
 import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class LocalStatisticsConfigStorage @Inject constructor(
-    @AppContext val context: Context,
-    @BaseGson private val gson: Gson,
+    @AppContext private val context: Context,
+    @BaseGson private val baseGson: Gson,
 ) {
     private val prefs by lazy {
         context.getSharedPreferences("statistics_local_config", Context.MODE_PRIVATE)
     }
 
-    val activeDistricts = prefs.createFlowPreference(
-        key = PKEY_ACTIVE_DISTRICTS,
-        reader = FlowPreference.gsonReader(gson, emptySet<SelectedDistrict>()),
+    private val gson by lazy {
+        baseGson
+            .newBuilder()
+            .registerTypeAdapterFactory(
+                RuntimeTypeAdapterFactory.of(SelectedStatisticsLocation::class.java)
+                    .registerSubtype(SelectedStatisticsLocation.SelectedDistrict::class.java)
+                    .registerSubtype(SelectedStatisticsLocation.SelectedFederalState::class.java)
+            )
+            .create()
+    }
```

Comment:
Please add a unit test for the serialization, sub-typing can be quite the footgun.

Reply:
oh it gunned me in the foot alright 😄